Improves the  wide flamer nozzle and related refactor

Lumipharon opened the pull request:
<!-- ***STOP!***  Read this: If this is not a PR ready for review and merge or WIP, open it as a draft PR, using the arrow next to 'Create Pull Request'>

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Edit: I have no reduced the flame duration of the wide nozzle by 20%, which means the wide nozzle will no longer be able to kill walls in a single click. You can still just click twice in a row to kill them, but it means you'll absolutely blast through fuel like nothing else.
I think making any flamer be able to kill walls in one click was a questionable change, but I'll reserve changing overall flamer behavior to another pr as that's out of scope.

So this was a rabbit hole of code to rework, guh.

Changed the way the wide nozzle calculates the turfs to burn, giving it more of a natural icecream cone sort of shape instead of a wide pyramid.

https://user-images.githubusercontent.com/7869430/187071890-a785afd5-f6ee-426f-9cbc-9da550b455fa.mp4

This should hopefully be actually useful, compared to the current iteration which is frankly niche if not outright bad.

Also refactors a bunch of flamer code, and some global procs related to CONES (literally the worst), and calculating valid pathing between turfs.
Notably there is now a cone proc that actually works (so for an obvious example prae acid cone could be changed as well, among other things), and flamers will now actually burn turfs that have border objects on them like windoors or barricades where it makes sense for them to do so.

I've tested this pretty thoroughly, but probs TM it anyway. Also maybe people will think that the range or angle of the cone should be tweaked.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Wurf flammen betterer
Bug fix good
I have finally conquered the vicious cone
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
balance: Changed the AOE calculation for the wide flamer nozzle to be more useful
fix: fixed flamers not igniting tiles with border objects like barricades or windoors in suitable cases
refactor: refactored some flamer and cone code
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

QualityVan commented on code/__HELPERS/unsorted.dm:
```diff
@@ -160,64 +160,59 @@ GLOBAL_REAL_VAR(list/stack_trace_storage)
+ *	projectile - whether throwpass will be checked to ignore dense objects like projectiles
+ *	bypass_xeno - whether to bypass dense xeno structures like flamers
+ */
+/proc/LinkBlocked(turf/A, turf/B, bypass_window = FALSE, projectile = FALSE, bypass_xeno = FALSE)
```
Seems like you could just have a PASSX bitflag here instead of running a separate var for each thing to check.

Lumipharon replied:
mesmoothbrain, what's that?

QualityVan replied:
Just a single var to use any combination of https://github.com/tgstation/TerraGov-Marine-Corps/blob/master/code/__DEFINES/equipment.dm#L1
You'd need to add a define for xeno walls, not 100% sure if any apply perfectly to what you want with bullets.

QualityVan commented on code/modules/projectiles/guns/flamer.dm:
```diff
@@ -69,6 +69,8 @@
 	///Damage multiplier for mobs caught in the initial stream of fire.
 	var/mob_flame_damage_mod = 2
+	///how wide of a cone the flamethrower produces on wide mode.
+	var/cone_angle = 55
```
Should this be on the nozzle itself rather than the gun?

Lumipharon replied:
There's no other nozzles that change it, and if there was then it would make sense for it to be a modifier to the base gun's angle.

BraveMole replied:
I kinda agree with QV here

Lumipharon replied:
The reason why I want to have it on the gun is so that I can add one or more new flamers in the future that different angles, while still being compatible with the sam nozzles.

QualityVan commented:
Around 9-10 seconds in the demo vid, why do some flames appear directly under you when you aren't firing?

Lumipharon commented:
> Around 9-10 seconds in the demo vid, why do some flames appear directly under you when you aren't firing?

Recursive flames is checking for tiles already flamed, relative to where you are. I assumed it's because you're holding down the trigger for a moment. It's not actually the tile I'm on, it's the tile in front of me as soon as I move.
This is intended, but if people think it's fucky it would be easy enough to use the location of src at the time of firing.

blackdav123 commented:
cone flamer being a niche tool that is generally bad was what it is designed for

if xenos are walling off entire sections of cave then you use cone flamer to stop that, this would make it the best maze clearing option in all scenarios

BraveMole commented:
> > Around 9-10 seconds in the demo vid, why do some flames appear directly under you when you aren't firing?
> 
> Recursive flames is checking for tiles already flamed, relative to where you are. I assumed it's because you're holding down the trigger for a moment. It's not actually the tile I'm on, it's the tile in front of me as soon as I move. This is intended, but if people think it's fucky it would be easy enough to use the location of src at the time of firing.

Look funky as hell yes

Lumipharon commented:
Updated to only run the code proc once and instead burn the tile in them layer by layer, so it should be a bit more efficient without changing it's behavior.

Also changed it to use get_turf(src) when first fired, so the cone won't shift if you move/get teleported etc.

TiviPlus commented on code/__HELPERS/unsorted.dm:
```diff
@@ -1210,4 +1205,57 @@ will handle it, but:
+	for(var/r in 1 to max_row_count)
+		if(r > 2)
+			cardinals = GLOB.cardinals
+		for(var/X in turfs_to_check) //checks the inital turf, then afterwards checks every turf that is added to cone_turfs
+			var/turf/trf = X
+			for(var/direction in cardinals)
+				var/turf/T = get_step(trf, direction) //checks all turfs around X
```
use AS, and please use proper var names this reads like algebra

Lumipharon replied:
done

TiviPlus reviewed (changes requested):
Rest lgtm I spose

BraveMole commented on code/__HELPERS/unsorted.dm:
```diff
@@ -1210,4 +1205,57 @@ will handle it, but:
+	if(left_angle < 0)
+		left_angle += 360
+
+	var/list/cardinals = GLOB.alldirs
```
is that a copy ?

BraveMole commented on code/__HELPERS/unsorted.dm:
```diff
@@ -1210,4 +1205,57 @@ will handle it, but:
+					continue
+				if(turf_angle > right_angle && turf_angle < left_angle)
+					continue
+				if(blocked)
+					if(LinkBlocked(trf, T, bypass_window, projectile, bypass_xeno))
```
only one if

Lumipharon replied:
whoops, used to be more there.

BraveMole commented on code/__HELPERS/unsorted.dm:
```diff
@@ -1210,4 +1205,57 @@ will handle it, but:
+				cone_turfs += T
+				turfs_to_check += T
+			turfs_to_check -= trf
+		for(var/turf/checked_turf in cone_turfs)
```
AS

Lumipharon replied:
done

BraveMole commented on code/modules/projectiles/guns/flamer.dm:
```diff
@@ -150,80 +152,59 @@
+	addtimer(CALLBACK(src, .proc/recursive_flame_cone, iteration, turfs_to_ignite, dir_to_target, range, current_target, flame_source), flame_spread_time)
 
 ///Checks and lights the turfs in turfs_to_burn
 /obj/item/weapon/gun/flamer/proc/burn_list(list/turf/turfs_to_burn)
```
? looks useless

BraveMole commented on code/modules/projectiles/guns/flamer.dm:
```diff
@@ -150,80 +152,59 @@
 		return
 
+	var/list/turf/turfs_by_iteration = list()
+	for(var/turf/turf in turfs_to_ignite)
```
AS

Lumipharon replied:
done

Lumipharon commented:
ahhh sorry, Was clicking the one button and it was just your names changing place.

Araulius commented:
this makes the standard nozzle obsolete, I hope you know that

Lumipharon commented:
> this makes the standard nozzle obsolete, I hope you know that

It absolutely chews through ammo like nothing. I burned through an entire backtank in like 20 minutes and I wasn't even spamming it as much as I would have liked.

Sure a single blast covers more area than the narrow nozzle, but that's not any more damage on anything you hit, and you can't choose to cover a shorter distance when that's not necessary. It's full blast or nothing.

Araulius commented:
> > this makes the standard nozzle obsolete, I hope you know that
> 
> It absolutely chews through ammo like nothing. I burned through an entire backtank in like 20 minutes and I wasn't even spamming it as much as I would have liked.
> 
> Sure a single blast covers more area than the narrow nozzle, but that's not any more damage on anything you hit, and you can't choose to cover a shorter distance when that's not necessary. It's full blast or nothing.

yes, but generally if you're a pyro nerd, you're taking point, and you know to bring welding tanks down to the planet. x fuel is gonna become a bajillion times more oppressive. hivelords are going to mald so hard.

Lumipharon commented:
> > > this makes the standard nozzle obsolete, I hope you know that
> > 
> > 
> > It absolutely chews through ammo like nothing. I burned through an entire backtank in like 20 minutes and I wasn't even spamming it as much as I would have liked.
> > Sure a single blast covers more area than the narrow nozzle, but that's not any more damage on anything you hit, and you can't choose to cover a shorter distance when that's not necessary. It's full blast or nothing.
> 
> yes, but generally if you're a pyro nerd, you're taking point, and you know to bring welding tanks down to the planet. x fuel is gonna become a bajillion times more oppressive. hivelords are going to mald so hard.

you can fire 25 times before a backtank is empty, you're gonna burn through that xfuel real quick

lKiseki commented:
As per this PR, the wide nozzle is a bit too oppressive. I have attached a screenshot of its range, which is worse than it's ever been.
![image](https://user-images.githubusercontent.com/59634950/189256092-8f2964ff-96af-48d9-b4f3-7ccd550f55e1.png)

For further reference, this is the standard nozzle flamer.
![image](https://user-images.githubusercontent.com/59634950/189256158-2dd41405-3faa-46d8-9a25-f21e70bc3fb0.png)

The other nozzle is just a fireball and one tile of fire, so I've decided to omit that.

This is objectively just a direct upgrade (and buff) to the flamer, allowing it to occupy a lot more range than what is healthy for balance. It is likewise complicated to deal with, as it chews through resin structures relatively quickly and leaves little space for xenos to maneuver. Gets worse with X fuel.
Fuel is not exactly an issue when you can bring plenty, and effectively invalidate the amount of fuel that is used up for every flamer shot.

QualityVan commented:
A: Why'd you nerf line flamer, B: why isn't it in the changelogs
